pkg/services/queryhistory/database.go

```diff
@@ -11,34 +11,65 @@ import (
 	"github.com/grafana/grafana/pkg/util"
 )
 
+// deletes runtime unique properties from the queries (such as `key` and `refId`)
+// so that we can compare queries
+func pruneExtraQueryProperties(queries []map[string]interface{}) []map[string]interface{} {
+	prunedQueries := make([]map[string]interface{}, 0)
+	for _, query := range queries {
+		delete(query, "key")
+		delete(query, "refId")
+		prunedQueries = append(prunedQueries, query)
+	}
+	return queries
+}
+
 // createQuery adds a query into query history
 func (s QueryHistoryService) createQuery(ctx context.Context, user *user.SignedInUser, cmd CreateQueryInQueryHistoryCommand) (QueryHistoryDTO, error) {
 	queryHistory := QueryHistory{
 		OrgID:         user.OrgID,
-		UID:           util.GenerateShortUID(),
-		Queries:       cmd.Queries,
 		DatasourceUID: cmd.DatasourceUID,
 		CreatedBy:     user.UserID,
-		CreatedAt:     time.Now().Unix(),
-		Comment:       "",
 	}
+	now := time.Now().Unix()
+
+	queryHistory.Queries = pruneExtraQueryProperties(cmd.Queries)
```

Comment:
existing queries will be eventually phased out (except for starred queries), but this makes probably more sense as it will also handle that situation. will give it a shot! 👍

Reply:
soooo, i've been thinking about that a bit and there are 3 possible ways of achieving this:

1. running it in the cleanup job (your suggestion)
   There are 2 downsides: the smallest one is that it should run only once, while the cleanup job runs every X hours. 
   this could be potentially avoided by using either the KV store and check whether this operation was performed or 
   not, or just let it run every time (the query selecting the data to modify should return 0 rows anyway after the first 
   run, provided we don't want to batch operations, in which case this could be technically considered a pro).
   the bigger one one is that the cleanup job doesn't run on startup, which means users will have a potentially half 
   cleaned up db while the cleanup catches up, which would probably make for a kinda bad experience.
2. running it on service initialization
  the downside here is that, while this would run when grafana starts, restarting grafana would run it again (same as previous point, we could use a kv store to check for this or just let the select query return 0 rows, but without potentially batching updates)
4. running it while migrating
  this really runs only once, but i'm not sure wether cleaning up existing data belongs here.

a note on "executing the select query" multiple times: while it's true that by returning 0 rows it would skip subsequent cleanup actions, the select itself may be expensive given it'd have to operate on potentially tens of thousands of non-indexed records. (we may consider adding an index to that field, which may benefit lookups when checking for queries existence as well, but that may need some benchmarking)

Reply:
I personally wouldn't mind if we didn't prune old entries as they will be removed in 14 days anyway. And basically they don't have any negative effect, besides that they won't be updated. 

If we really wanted to prune, I'd personally only prune starred queries, not all queries in db. Seems like overkill to me.

Reply:
yeah, probably only pruning starred queries makes much more sense as they'll be in fact the only ones to be really affected

Reply:
A tale about this:
implemented it in [19bdaad](https://github.com/grafana/grafana/pull/63627/commits/19bdaada018a67193e3cc6f47c8d7b7b374fee3f)

it runs at service startup when query history is enabled in a goroutine in order to avoid blocking grafana's startup if the pruning takes too long. additionally, the pruning is done in batch of 500, which should be a reasonable number (also considering we are not expecting a huge amount of starred queries to be present in any instance).

While testing i noticed that even when batching updates, only a single entry was being updated in the DB. I traced back this to https://github.com/mattn/go-sqlite3/issues/933, and it seems that multiple statements are not really supported by go-sqlite3 (which we use as a transient dependency of xorm 0.8.2).
On a side note, also sqlx (the library we are eventually going to switch to) uses go-sqlite3, so waiting for the switch before implementig this is not an option.

I ended up removing the batching in [9b53864](https://github.com/grafana/grafana/pull/63627/commits/9b538642314afdb8d410c9b8151ba0f1c2951570), but without it i'm very dubious about updating queries in the first place. the current star limit is `150000`, assiming the worst possible case, it means that running this would require 150000 queries to be executed at startup, and while it's true that they'd be executed only once, i'm not very comfortable with it 🤔